### PR TITLE
ENH: special: JAX support (non-jitted)

### PR DESCRIPTION
**Pull request by crusaderky**
- Part of #22246
- See also #22255

### CODE REVIEW

File: scipy/special/_logsumexp.py [hunk 1]
```diff
@@ -144,11 +144,10 @@ def logsumexp(a, axis=None, b=None, keepdims=False, return_sign=False):
 def _wrap_radians(x, xp=None):
     xp = array_namespace(x) if xp is None else xp
     # Wrap radians to (-pi, pi] interval
-    out = -((-x + math.pi) % (2 * math.pi) - math.pi)
-    # preserve relative precision
+    wrapped = -((-x + xp.pi) % (2 * xp.pi) - xp.pi)
+    # Preserve relative precision
```
Comment (crusaderky) [inline scipy/special/_logsumexp.py:148]:
```suggestion
    # preserve relative precision
```

File: scipy/special/_logsumexp.py [hunk 2]
```diff
@@ -172,11 +171,10 @@ def _elements_and_indices_with_max_real(a, axis=-1, xp=None):
         # simple, array-API compatible way of doing so that doesn't
         # have a problem with `axis` being a tuple or None.
         i = xp.reshape(xp.arange(xp_size(a)), a.shape)
-        i[~mask] = -1
+        i = xpx.at(i, ~mask).set(-1)
```
Comment (crusaderky) [inline scipy/special/_logsumexp.py:174]: This fails on jax.jit. My current intention is to change jax.jit itself to special-case `arr.at[idx].set(value)` when idx is a boolean mask and value is a scalar, so that it can be rewritten as `jnp.where(idx, value, arr)`. Failing that, I can implement the same special case in array-api-extra.

File: scipy/special/tests/test_logsumexp.py [hunk 3]
```diff
@@ -172,12 +166,13 @@ def test_logsumexp_b_shape(self, xp):
...
     @pytest.mark.parametrize('arg', (1, [1, 2, 3]))
-    @pytest.mark.skip_xp_backends(np_only=True)
+    @pytest.mark.skip_xp_backends(np_only=True, reason="pure Python")
+    @pytest.mark.usefixtures("skip_xp_backends")
     def test_xp_invalid_input(self, arg, xp):
         assert logsumexp(arg) == logsumexp(np.asarray(np.atleast_1d(arg)))
 
-    @pytest.mark.skip_xp_backends(np_only=True,
-                                  reason="Lists correspond with NumPy backend")
+    @pytest.mark.skip_xp_backends(np_only=True, reason="pure Python")
```
Comment (lucascolley) [inline scipy/special/tests/test_logsumexp.py:174]: nit: I would explain this as "array-like input" rather than "pure Python"

File: scipy/special/_logsumexp.py [hunk 4]
```diff
@@ -144,11 +142,10 @@ def logsumexp(a, axis=None, b=None, keepdims=False, return_sign=False):
 def _wrap_radians(x, xp=None):
     xp = array_namespace(x) if xp is None else xp
     # Wrap radians to (-pi, pi] interval
-    out = -((-x + math.pi) % (2 * math.pi) - math.pi)
+    wrapped = -((-x + xp.pi) % (2 * xp.pi) - xp.pi)
     # preserve relative precision
     no_wrap = xp.abs(x) < xp.pi
-    out[no_wrap] = x[no_wrap]
-    return out
+    return xp.where(no_wrap, x, wrapped)
```
Comment (lucascolley) [inline scipy/special/_logsumexp.py:148]: for my understanding, could you explain why we use `where` here instead of `xpx.at`?

File: scipy/special/_logsumexp.py [hunk 5]
```diff
@@ -172,11 +169,10 @@ def _elements_and_indices_with_max_real(a, axis=-1, xp=None):
         # simple, array-API compatible way of doing so that doesn't
         # have a problem with `axis` being a tuple or None.
         i = xp.reshape(xp.arange(xp_size(a)), a.shape)
-        i[~mask] = -1
+        i = xpx.at(i, ~mask).set(-1)
         max_i = xp.max(i, axis=axis, keepdims=True)
         mask = i == max_i
-        a = xp_copy(a)
-        a[~mask] = 0
+        a = xpx.at(a, ~mask).set(0, copy=True)
```
Comment (mdhaber) [inline scipy/special/_logsumexp.py:175]: I'd like to stay current with this stuff even though I'm not working on array API ATM. Can you explain when `copy=True` would be preferred to just using `xp.where`? (ISTM this could/should have been written with `xp.where` initially, if it wasn't really modifying an existing array.)

File: scipy/special/_logsumexp.py (hunk 4 above)
Comment (crusaderky) [reply scipy/special/_logsumexp.py:148]: This is for forward compatibility with jitted jax, which doesn't support updates with bool masks. Performance implications are negligible in this case.

File: scipy/special/_logsumexp.py (hunk 5 above)
Comment (crusaderky) [reply scipy/special/_logsumexp.py:175]: `xpx.at(a, mask).set(value, copy=True)` differs from where when value is not a scalar.
I fully agree that xp.where is much more readable in this case; I'm updating it.

Comment (crusaderky) [reply scipy/special/_logsumexp.py:175]: There's a caveat regarding dtypes though:
```python
xpx.at(a, ~mask).set(0, copy=True)
```
retains the dtype of a, whereas
```python
xp.where(mask, a, 0)
```
broadcasts the dtypes. To retain the behaviour, you need 
```python
xp.where(mask, a, xp.zeros((), dtype=a.dtype))
```

Comment (lucascolley) [reply scipy/special/_logsumexp.py]: that should be resolved with https://github.com/data-apis/array-api/pull/860

File: scipy/special/_logsumexp.py (hunk 4 above)
Comment (jakevdp) [reply scipy/special/_logsumexp.py:148]: In general, it ensures compatibility with immutable arrays – that is to say, this is more important than just supporting jitted JAX, in my opionion.

Comment (mdhaber) [reply scipy/special/_logsumexp.py:148]: > In general, it ensures compatibility with immutable arrays

But isn't that what `at.set` is for?

File: scipy/special/_logsumexp.py (hunk 5 above)
Comment (mdhaber) [reply scipy/special/_logsumexp.py]: > There's a caveat regarding dtypes though:

Right, that's what I was referring to with "It might have been this way to avoid jumping through hoops to preserve dtype with `where`." I see that your update jumped through the hoops.

Looking forward to http://data-apis/array-api#860...

File: scipy/special/_logsumexp.py (hunk 4 above)
Comment (crusaderky) [reply scipy/special/_logsumexp.py:148]: > > In general, it ensures compatibility with immutable arrays
> 
> But isn't that what `at.set` is for?

Yes. In this case, the only beneficiary of xp.where vs. xpx.at is jitted jax.

File: scipy/special/_logsumexp.py (hunk 5 above)
Comment (crusaderky) [reply scipy/special/_logsumexp.py]: > > There's a caveat regarding dtypes though:
> 
> Right, that's what I was referring to with "It might have been this way to avoid jumping through hoops to preserve dtype with `where`." I see that your update jumped through the hoops.
> 
> Looking forward to http://data-apis/array-api#860...

I can revert it and add a comment explaning the choice if it looks more readable to you.

File: scipy/special/_logsumexp.py (hunk 4 above)
Comment (lucascolley) [reply scipy/special/_logsumexp.py:148]: > But isn't that what at.set is for?

I believe no, when the library does not support [the `boolean indexing` capability](https://data-apis.org/array-api/2023.12/API_specification/generated/array_api.info.capabilities.html#capabilities). If that's correct then we should add a note to https://data-apis.org/array-api-extra/generated/array_api_extra.at.html to make that clear.

Comment (mdhaber) [reply scipy/special/_logsumexp.py:148]: OK... and based on https://github.com/scipy/scipy/pull/22256/files#r1904169337, am I to understand that *all* of these currently fail with jitted jax? It looks like all have `idx` as a boolean mask and `value` a scalar.

Why is it important for *this instance* to be forward compatible with jitted jax and not all the rest?

Comment (lucascolley) [reply scipy/special/_logsumexp.py:148]: nevermind, that must be wrong:

```ipython
In [5]: jnp.__array_namespace_info__().capabilities()
Out[5]: {'boolean indexing': True, 'data-dependent shapes': False}
```

Comment (lucascolley) [reply scipy/special/_logsumexp.py:148]: > The concern with changing to xp.where now instead of leaving it broken is that there is no reminder to come back and improve it when there is support for value being an array.

I had the same thought and left a reminder at https://github.com/scipy/scipy/issues/22246#issuecomment-2573357797

**Comment by lucascolley** [comment]
> If this change is intended as a partial step toward eventual support for traced execution, then that's great, and we should do it.

yes: gh-22246

File: scipy/special/_logsumexp.py (hunk 5 above)
Comment (mdhaber) [reply scipy/special/_logsumexp.py]: Naw, it's fine as is. There are lots of times I've done this. They stick out to me as ugly, so I will notice them and clean them up when it's possible.

**Comment by crusaderky** [comment]
All CI failures are unrelated

File: scipy/special/_logsumexp.py (hunk 4 above)
Comment (rgommers) [reply scipy/special/_logsumexp.py:148]: Quick comment, since I think the discussion above misses a point: `xpx.at` _is_ compatible with `jax.jit` and immutable arrays. The change is because `xpx.at` currently _does not support_ `__setitem__` with a bool array as index, only integer array.

The more mechanical change can only happen after there's a like-for-like replacement for boolean `x[idx] = ...` (which is not `xp.where`).

Comment (crusaderky) [reply scipy/special/_logsumexp.py:148]: The reason to change this now is that in this specific case it is _trivial_ to do so.
When it comes to xpx.at + jax.jit + boolean mask indices, there are three substantially different use cases to tackle:

1. when the expression could be trivially rewritten as a `xp.where` with no performance loss (this case)
2. when the value is a scalar, so we could somewhat straightforwardly implement a special case rewrite rule either in jax.jit itself or inside xpx.at that converts it, for jax only, to `jnp.where(mask, scalar_value, x)`; after that change to jax/array_api_extra the changes I'm doing _now_ in scipy will automatically become compatible with jax.jit.
3. when neither is true, e.g. `_lazywhere` and more ad-hoc work is required

> And how are we deciding when the performance impact is negligible? 

Is there a benchmark suite we can run? I would be reluctant to run it manually for every PR though.

Comment (mdhaber) [reply scipy/special/_logsumexp.py:148]: @rgommers can you clarify your opinion here - should the mutation be replaced by `where` on this line, and if so, why is that right to do that here and not elsewhere?

You wrote "The change is because xpx.at currently does not support __setitem__ with a bool array as index, only integer array." but AFAICT,  https://github.com/scipy/scipy/pull/22256/files#r1904169337 also has a boolean array as the index. What is the difference between these two cases?

> Is there a benchmark suite we can run? I would be reluctant to run it manually for every PR though.

It doesn't [look like it](https://pv.github.io/scipy-bench/), at least there wasn't last time that ran. I wouldn't want to run it for every PR either; that's why I propose that we try to maintain mutation for mutable arrays rather than deciding whether or not there is a performance cost.

Comment (rgommers) [reply scipy/special/_logsumexp.py:148]: > @rgommers can you clarify your opinion here - should the mutation be replaced by `where` on this line, and if so, why is that right to do that here and not elsewhere?

I spent >5 minutes reading the surrounding code and while I suspect @crusaderky is right that it doesn't matter performance-wise, it's not completely obvious to me (`where` operands are arrays of the same size as the output, so it could slow things down by some %). So while I'm not attached to the final outcome for this particular case, I'd encourage fixing the design issue first so future `x[mask] = ...` instances can be replaced by something that is 100% semantically equivalent, so there's no need to think at all.

> If that's not possible with `xpx.at.set` as-written, then why not improve `xpx.at.set` before this sweep is done?

I agree with this. I don't have a strong opinion of whether it must be inside `at` or in another function, but I think it's probably better for it to be a single thing, rather than the three cases 1-3 above.

Comment (crusaderky) [reply scipy/special/_logsumexp.py:148]: > I agree with this. I don't have a strong opinion of whether it must be inside at or in another function, but I think it's probably better for it to be a single thing, rather than the three cases 1-3 above.

I don't see how the use case of `_lazywhere` could be made to work with a simple helper function, honestly.

As for use case 1, I _could_ write something like

```
class at:
    def set_masked(self, value, copy=None):
        """Same as x[idx] = value[idx]"""
```
but I think it would both be less readable _and_ slower compared to `x = xp.where(idx, value, x)` including on writeable backends. Benchmark here: https://github.com/scipy/scipy/issues/22246#issuecomment-2574205151

Comment (crusaderky) [reply scipy/special/_logsumexp.py:148]: >  but AFAICT, https://github.com/scipy/scipy/pull/22256/files#r1904169337 also has a boolean array as the index. What is the difference between these two cases?

The difference is that https://github.com/scipy/scipy/pull/22256/files#r1904169337 is
```python
a[mask] = scalar
```
which can be special-cased for JAX inside JAX itself or array-api-extra so that it works with the jit, and would be slower to rewrite in numpy as
```python
a = xp.where(mask, scalar, a)
```
benchmark here: https://github.com/scipy/scipy/issues/22246#issuecomment-2574218700

This one instead is
```python
a[mask] = b[mask]
```
which _cannot_ be special-cased transparently; we could write a new method `a = at(a)[mask].set_masked(b)` which not only it would hurt readability, but it would also be at all times slower on numpy than just writing
```python
a = xp.where(mask, b, a)
```
benchmark here: https://github.com/scipy/scipy/issues/22246#issuecomment-2574205151

**Review by lucascolley — APPROVED** [review]
thanks @crusaderky !

**Comment by lucascolley** [comment]
@crusaderky I noticed that this skip remains:

https://github.com/scipy/scipy/blob/9166b6fc4714fa55d4e829c8a76a278b8ecaf647/scipy/special/tests/test_support_alternative_backends.py#L68-L69

Perhaps that could/should have been addressed here?

**Comment by crusaderky** [comment]
> @crusaderky I noticed that this skip remains:
> 
> https://github.com/scipy/scipy/blob/9166b6fc4714fa55d4e829c8a76a278b8ecaf647/scipy/special/tests/test_support_alternative_backends.py#L68-L69
> 
> Perhaps that could/should have been addressed here?

https://github.com/scipy/scipy/pull/22281
